File: leetcode/peak-index-in-a-mountain-array.py
from typing import List


class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        l = 0
        r = len(arr)

        while l < r:
            m = (l + r) // 2

            if arr[m] > arr[m - 1]:
                # upward incline
                l = m + 1
            else:
                r = m

        return l - 1

File: leetcode/peak-index-in-a-mountain-array.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    int peakIndexInMountainArray(vector<int>& arr)
    {
        auto l = 0;
        auto r = arr.size() - 1;

        while (l < r) {
            auto m = l + (r - l) / 2;

            if (arr[m] < arr[m + 1]) {
                l = m + 1;
            } else {
                r = m;
            }
        }

        return l;
    }
};
